Pre 1.3

cchalmers opened the pull request:
This is a bunch of miscellaneous changes I wanted to make for 1.3. I've tried to make each commit a separate feature. The main changes are:

##### Specialise `Colour` and `AlphaColour` `Color` instances to `Double`
This means you can write `fillColour black` without a type signature on `black`. If you happen to be working with a `Colour Float` you can use `colourConvert`.

##### Move (`|>`) (qualify object with `Name`) to (`>|`)
This cleans up space for lenses snoc operator to use with `Line`s new `Cons` and `Snoc` instances.

##### Remove `LineJoinA` and `LineCapA`
I'm not too bothered about this, it just seemed silly to make a `newtype` around `Last` just for the instance.

##### Move `view` to `rectEnvelope`
I'm open to name suggestions but I use `view` a lot from lens and I always forget and get horrible error messages. Also I don't think `view`s a great name because anything outside the box is still visible, it just happens to hide it if you use it on the final diagram.

##### New show instances
The new show instances hide the internal representation and can be copy pasted to form valid Haskell when possible. `Attribute` and `Style` instances show the name of type they're holding which is useful for debugging

##### Export lens from prelude
It doesn't have to be the whole of lens but I'd like a lot more. Importing lens separately makes a few annoying clashing and it is a prelude after all :)

cchalmers commented:
This shouldn't cause many problems with backends. Some of them fail `-Werror` because of the extra lens exports.

byorgey commented on src/Diagrams/Prelude.hs:
```diff
@@ -132,8 +140,11 @@ module Diagrams.Prelude
 
-         -- | Essential Lens Combinators
-       , (&), (.~), (%~)
+         -- | This exports most of the lens module, excluding the
```
Should be lens "package" not module.  Or maybe it should say "Control.Lens module".

byorgey commented:
My responses to your main changes are below.  I'm still not done looking carefully through all the changes.

##### Specialise `Colour` and `AlphaColour` `Color` instances to `Double`
Makes sense.

##### Move (`|>`) (qualify object with `Name`) to (`>|`)
Also makes sense, though as expressed at https://github.com/diagrams/diagrams-core/pull/82 I am not the biggest fan of the name (`>|`).

##### Remove `LineJoinA` and `LineCapA`
I'm a pretty big fan of type-directed instances.  But in this case perhaps you are right.

##### Move `view` to `rectEnvelope`
Indeed, `view` is a bad name and it would be good not to conflict with such a basic function from `lens`.  I think `rectEnvelope` sounds like a good name.

##### New show instances
Cool.

##### Export lens from prelude
Makes sense to me.

cchalmers commented:
I forgot to mention 72bf032 -- transformation isomorphisms.

I've added `transformed`, `movedTo`, `movedFrom`, `translated` and `rotated` `Iso`s and moved `under` to `underT`.  The new isomorphisms can be used with `lens`'s `under`. So `rotateAround` could be implemented as:

```
rotateAround p angle = under (movedFrom p) (rotate angle)
```

I've added example usage in the comments for the new `Iso`s.

bergey commented:
I've wanted those transformation `Iso`s for a while.  Also the proper `Show` instance for `Angle`s.

byorgey commented:
Ah, I like it. I had noticed some code using movedFrom like that and
wondered what was going on. We should make sure to update the part of the
user manual that used to talk about under to mention these.

On 9:48am, Tue, Mar 10, 2015 Chris <notifications@github.com> wrote:

> I forgot to mention 72bf032
> <https://github.com/diagrams/diagrams-lib/commit/72bf032bcef0e740877297676950229ea6eb79da>
> -- transformation isomorphisms.
>
> I've added transformed, movedTo, movedFrom, translated and rotated Isos
> and moved under to underT. The new isomorphisms can be used with lens's
> under. So rotateAround could be implemented as:
>
> rotateAround p angle = under (movedFrom p) (rotate angle)
>
> I've added example usage in the comments for the new Isos.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/diagrams/diagrams-lib/pull/241#issuecomment-78055617>.
>

byorgey commented:
Note that `lens-4.8` is out now, so we should add support for that as well.  I don't know whether we need to support lens < 4.6; probably not.

byorgey commented:
So what's still holding this up?  Just deciding what to do with the `lens` re-exports?

byorgey commented:
Oh, wait, I see those changes have already been made.  This looks good to me.

jeffreyrosenbluth commented:
LGTM as well

bergey commented:
Great!
